lib/AST/USRGeneration.cpp

```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
   return false;
 }
 
+static bool printObjCNameFragment(const ValueDecl *D, StringRef ObjCName,
+                                  raw_ostream &OS) {
+  if (!D)
+    return true;
+
+  if (isa<ClassDecl>(D)) {
+    clang::index::generateUSRForObjCClass(ObjCName, OS);
+  } else if (isa<ProtocolDecl>(D)) {
+    clang::index::generateUSRForObjCProtocol(ObjCName, OS);
+  } else if (isa<VarDecl>(D)) {
+    clang::index::generateUSRForObjCProperty(ObjCName, D->isStatic(), OS);
+  } else if (isa<AbstractFunctionDecl>(D)) {
+    clang::index::generateUSRForObjCMethod(ObjCName, D->isInstanceMember(), OS);
+  } else if (isa<EnumDecl>(D)) {
+    OS << "@E@" << ObjCName; // FIXME: expose clang API to handle enum names
+  } else if (isa<EnumElementDecl>(D)) {
+    OS << "@" << ObjCName;
+  } else {
+    llvm_unreachable("Unexpected value decl");
+    return true;
+  }
+  return false;
+}
+
+static bool printObjCUSR(const ValueDecl *D, raw_ostream &OS, Identifier Ident,
+                         ObjCSelector Selector) {
+  OS << clang::index::getUSRSpacePrefix();
+
+  if (D->getDeclContext()->isTypeContext()) {
+    auto *Parent = D->getDeclContext()->getAsNominalTypeOrNominalTypeExtensionContext();
+    auto ObjCName = objc_translation::getObjCNameForSwiftDecl(Parent);
+    if (printObjCNameFragment(Parent, ObjCName.first.str(), OS))
+      return true;
+  }
+
+  if (!Ident.empty())
+    return printObjCNameFragment(D, Ident.str(), OS);
+  if (Selector) {
```

Comment:
I realize we do have similar forgiving check in `SwiftSourceDocInfo.cpp`. The reason there is that we may have preferred names that are non-translatable to ObjC names, for instance, when the argument counts do not match. However, these cases are unlikely to happen here because we don't have preferred names.